keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,927 @@
+     a. As the scheduler keeps a tab on the available resources of the node, post updating the available capacities,
+        the scheduler proceeds to schedule any pending pods.
+
+
+#### Flow Control for updating swap limit for containers
+
+Formula to calculate Swap Limit: `(<containerMemoryRequest>/<nodeTotalMemory>)*<totalPodsSwapAvailable>`
+```
+T=0: Node Resources:
+        - Memory: 6G
+        - Swap: 4G
+     Pod:
+        - container1
+            - MemoryRequest: 2G
+            - State: Running
+     Runtime:
+        - <cgroup_path>/memory.swap.max: 1.33G
+        
+T=1: Resize Instance to Hotplug Memory:
+        - Memory: 8G
+        - Swap: 4G
+     Pod:
+        - container1
+            - MemoryRequest: 2G
+            - State: Running
+     Runtime:
+        - <cgroup_path>/memory.swap.max: 1G
+```
+
+Similar flow is applicable for updating oom_score_adj.
+
+### Handling HotUnplug Events
+
+Though this KEP focuses only on resource hotplug, It will enable the kubelet to capture the current available capacity of the node (Regardless of whether it was a hotplug or hotunplug of resources.)
+For now, we will introduce an error mode in the kubelet to inform users about the shrink in the available resources in case of hotunplug.
+
+As the hot-unplug events are not completely handled in this KEP, in such cases, it is imperative to move the node to the NotReady state when the current capacity of the node
+is lesser than the initial capacity of the node. This is only to point at the fact that the resources have shrunk on the node and may need attention/intervention.
+
+Once the node has transitioned to the NotReady state, it will be reverted to the ReadyState once when the node's capacity is reconfigured to match or exceed the last valid configuration.
+In this case, valid configuration refers to a state which can either be previous hot-plug capacity or the initial capacity in case there was no history of hotplug.
+
+#### Flow Control
+
+```
+T=0: Node initial Resources:
+    - Memory: 10G
+    - Pod: Memory
+
+T=1: Resize Instance to Hotplug Memory
+    - Current Memory: 10G
+    - Update Memory: 15G
+    - Node state: Ready
+
+T=2: Resize Instance to HotUnplug Memory
+    - Current Memory: 15G
+    - UpdatedMemory: 5G
+    - Node state: NotReady
+
+T=3: Resize Instance to Hotplug Memory
+    - Current Memory: 5G
+    - Updated Memory Size: 15G
+    - Node state: Ready
+```
+
+Few of the concerns surrounding hotunplug are listed below
+* Pod re-admission:
+    * Given that there is probability that the current Pod resource usage may exceed the available capacity of node, its necessary to check if the pod can continue Running
+      or if it has to be terminated due to resource crunch.
+* Recalculate OOM adjust score and Swap limits:
+    * Since the total capacity of the node has changed, values associated with the nodes memory capacity must be recomputed.
+* Handling unplug of reserved CPUs.
```

Comment:
Do we capture the corner case where node capacity would increase but some of the resources go away, e.g. new cpuset is bigger but doesn't contain the old cpuset (dumb example `0-3` -> `2-9`). CPU manager sync returns an error and node goes to not ready?

Reply:
So for now we thought to only record an event during resync failures. may be need to check how we can identify what was the assigned cpu set before hotplug.